src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs

```diff
@@ -151,31 +164,27 @@ public void Serialize(ObjectWriter writer, SolutionReplicationContext context, C
 
         var checksumAlgorithm = (SourceHashAlgorithm)reader.ReadInt32();
         var encoding = reader.ReadEncoding();
+        var contentHash = ImmutableCollectionsMarshal.AsImmutableArray(reader.ReadByteArray());
 
         var kind = (SerializationKinds)reader.ReadInt32();
         Contract.ThrowIfFalse(kind is SerializationKinds.Bits or SerializationKinds.MemoryMapFile);
 
         if (kind == SerializationKinds.MemoryMapFile)
         {
-            var storage2 = (ITemporaryStorageService2)storageService;
+            var storage2 = (TemporaryStorageService)storageService;
 
             var name = reader.ReadRequiredString();
             var offset = reader.ReadInt64();
             var size = reader.ReadInt64();
 
-            var storage = storage2.AttachTemporaryTextStorage(name, offset, size, checksumAlgorithm, encoding);
-            if (storage is ITemporaryTextStorageWithName storageWithName)
-            {
-                return new SerializableSourceText(storageWithName);
-            }
-            else
-            {
-                return new SerializableSourceText(storage.ReadText(cancellationToken));
-            }
+            var storage = storage2.AttachTemporaryTextStorage(name, offset, size, checksumAlgorithm, encoding, contentHash);
```

Comment:
by being strongly typed, we know we know we'll always get a ITemporaryTextStorageWithName back.